errors/errors.go

```diff
@@ -0,0 +1,212 @@
+	if e != nil && e.err != nil {
+		return e.err.Error()
+	}
+	return ""
+}
+
+// Unwrap implements the error unwrapping interface.
+func (e *Error) Unwrap() error {
+	return e.err
+}
+
+// Description returns the description of the error.
+func (e *Error) Description() string {
+	if e.description != "" {
+		return e.description
+	}
+	return e.err.Error()
+}
+
+// WithErrorReason used to pass reason, httpCode, and
+// grpcStatus code to the Error struct.
+func WithErrorReason(reason string, httpCode int, grpcStatusCode codes.Code) Option {
+	return func(err *Error) {
+		err.reason = reason
+		err.grpcStatusCode = grpcStatusCode
+		err.httpCode = httpCode
+	}
+}
+
+// WithResourceInfo used to pass ResourceInfo to the Error struct.
+func WithResourceInfo(resourceInfo *ResourceInfo) Option {
+	return func(e *Error) {
+		e.resourceInfo = resourceInfo
+	}
+}
+
+// WithDescription used to pass a description
+// to the Error struct.
+func WithDescription(description string) Option {
+	return func(e *Error) {
+		e.description = description
+	}
+}
+
+// WithMetadata used to pass a Metadata[string]string
+// to the Error struct.
+func WithMetadata(md map[string]string) Option {
+	return func(e *Error) {
+		e.metadata = md
+	}
+}
+
+func newErrorInfo(reason string, md map[string]string) *errdetails.ErrorInfo {
+	return &errdetails.ErrorInfo{
+		Domain:   Domain,
+		Reason:   reason,
+		Metadata: md,
+	}
+}
+
+func newResourceInfo(rid *ResourceInfo, err error) *errdetails.ResourceInfo {
+	return &errdetails.ResourceInfo{
+		ResourceType: rid.Type,
+		ResourceName: rid.Name,
+		Owner:        Owner,
+		Description:  err.Error(),
+	}
+}
+
+// *** GRPC Methods ***
+
+// GRPCStatus returns the gRPC status.Status object.
+func (e *Error) GRPCStatus() *status.Status {
+	messages := []protoiface.MessageV1{
+		newErrorInfo(e.reason, e.metadata),
+	}
+
+	if e.resourceInfo != nil {
+		messages = append(messages, newResourceInfo(e.resourceInfo, e.err))
+	}
+
+	ste, stErr := status.New(e.grpcStatusCode, e.description).WithDetails(messages...)
+	if stErr != nil {
+		return status.New(e.grpcStatusCode, e.description)
+	}
+
+	return ste
+}
+
+// *** HTTP Methods ***
+
+// ToHTTP transforms the supplied error into
+// a GRPC Status and then Marshals it to JSON.
+// It assumes if the supplied error is of type Error.
+// Otherwise, returns the original error.
+func (e *Error) ToHTTP() (int, []byte) {
+	resp, err := protojson.Marshal(e.GRPCStatus().Proto())
+	if err != nil {
+		return http.StatusInternalServerError, []byte(err.Error())
+	}
+
+	return e.httpCode, resp
+}
+
+// HTTPCode returns the value of the HTTPCode property.
+func (e *Error) HTTPCode() int {
+	if e.httpCode == 0 {
+		return http.StatusInternalServerError
+	}
+	return e.httpCode
+}
+
+// JSONErrorValue implements the errorResponseValue interface.
+func (e *Error) JSONErrorValue() []byte {
+	b, err := protojson.Marshal(e.GRPCStatus().Proto())
+	if err != nil {
+		return []byte(err.Error())
```

Comment:
This is not going to work because `err.Error()` is not a JSON-encoded string.

```suggestion
        errJSON, _ := json.Marshal(fmt.Sprintf("failed to encode proto to JSON: %v", err))
		return errJSON
```